compatibility_server/pip_checker.py

```diff
@@ -353,24 +356,44 @@ def _build_command(self, subcommands: List[str]):
 
     def _install(self, container: docker.models.containers.Container):
         """Run pip install in the container."""
-        command = self._build_command(['install', '-U'] + self._packages)
-        returncode, output = self._run_command(
-            container,
-            command,
-            stdout=False,
-            stderr=True,
-            raise_on_failure=False)
-        if returncode:
-            # Checking for environment error
-            environment_error = PIP_ENVIRONMENT_ERROR_PATTERN.search(output)
-            if environment_error:
-                raise PipError(error_msg=environment_error.group('error'),
-                               command=command,
-                               returncode=returncode)
-
-            return PipCheckResult(self._packages,
-                                  PipCheckResultType.INSTALL_ERROR,
-                                  output)
+        # If there is cloud client libraries, install it from source.
+        # Else install it from PyPI.
+        for pkg in self._packages:
```

Comment:
I'd structure this a bit differently to make it more readable.

```python

def _clone_repo(self, container, repo):
 """<docstring>"""
  _, directory = self._run_command(
    container,
    "echo $(mktemp -d)"
    stderr=False,
    stdout=True,
    raise_on_failure=True)

  shallow_clone_command = ['git', 'clone', '--depth', '1', repo, directory]
  self._run_command(
    container,
    shallow_clone_command,
    stdout=False,
    stderr=True,
    raise_on_failure=True)
  return directory

def _install(self, container: docker.models.containers.Container):
if any(GITHUB_CLIENTLIBS_PREFIX in pkg for pkg in self._packages):
  # The https://github.com/googleapis/google-cloud-python repository is large,
  # has a large amount of history and contains many independent packages.
  # Installing packages from that repository is quite slow e.g.
  # $ time pip install git+git://...
  # <result>
  # As an optimization, if any provided package is installed from this repository,
  # it is first cloned (using --depth=1 to avoid unnecessary history) and then
  # installed from local disk.
  client_repo_directory = self._clone_repo(CLIENTLIBS_GITHUB_URL)

install_names = []
for pkg in self._packages:
  if GITHUB_CLIENTLIBS_PREFIX in pkg:
    install_subdirectory = pkg.split(GITHUB_CLIENTLIBS_PREFIX)[1]
    install_names.append(os.path.join(client_repo_directory, install_subdirectory))
  else:
    install_names.append(pkg)
command = self._build_command(['install', '-U'] + install_names)
# Rest is the same.
```

Reply:
Done.